### DISCUSSION TITLE
XmlSerializer & Visitors

### DISCUSSION THREAD

**Discussion by cybercop23** [xLightsSequencer/xLights#4526] [General]
Per discussion, starting work on changing the XML export/import to the new XmlSerializer (should it be Serialiser?) and Visitors support.

Took the original work for base model and MHAdv and added the rest of the parameters for the default models (WIP).
We'll run though a few on our own RGBEffects files to make sure no parms are missing.

@AzGilrock  is this what you are loooking for - please say yes.. took me like an hour :)
Also, I haven't looked into, but assuming the vars can be anything at this point vs the actual variables used in the code...
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/208cfabc-e9ba-4215-9c10-84045ccee3eb)
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/3d0f739e-55ea-4db1-9e59-51d4ecb48a28)

**Comment by AzGilrock** [comment]
Yes that looks like the general idea.  Before anything gets commited I need to double-check it against the branch I already have in progress that has already updated that file.   I had already created some of that stuff.  It gets hard to have multiple people trying to make these changes.  One of the biggest hurdles is adding all the missing variables to the model class objects and having getters and setters for the XmlSerializer to retrieve and set the data.

Regarding naming....I named it first so we get the American spelling. :)
https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer?view=net-8.0

**Reply by cybercop23** [reply]
Yeah.. it gets hard when several folks work on the same part. If you want to check in what you have.. I can try to figure out how to merge it and then I can do teh rest of the grunt work (that's all I'm good at). From my understanding we can/should be able to just merge it in main as we add things, so long nothing uses it, we shoudl be  fine. Then it may be easier to manage updates.

**Comment by cybercop23** [comment]
https://github.com/xLightsSequencer/xLights/pull/4521#issuecomment-2076255991

**Reply by cybercop23** [reply]
https://github.com/xLightsSequencer/xLights/pull/4521#issuecomment-2076342465

**Reply by AzGilrock** [reply]
You said we were crossed but I was reading everything and thought we were fine.  You told me to check in what I have so now I was just waiting for you to merge into that.

Our first step can be to get all the import/export stuff working.  The next part is harder which will be starting to remove everything that relies on ModelXML.  So like inside every model where it handles parameter changes we should be able to switch the code to just assign the model's own class variable instead of doing a DeleteAttribute and AddAttribute operation.  I expect it to all be faster in the end because all these XML operations with strings can't be fast.  The import/export can be done one model at a time.  The next step I'm not sure can be done in pieces because once you stop updating the ModelXML object for a model then everything in the program that relies on retrieving model state from the XML will need to be updated.  I did a search on ModelXML and I think its in like 3,000+ locations.

**Reply by AzGilrock** [reply]
I have 2 other things I'm trying to knock out this weekend then I'll be back on this effort.  I'm working on adding a Dimmer tab to the MovingHead effect panel and I need to do something to deal with mapping the head locations during import.

I have a branch called "ban_xml" that I'll get pushed.  Its basically everything I checked in last night minus some of the parameter work in ArchesModel I had already started to work on I left out since that part needs to be completely done before pushing to master.

**Reply by cybercop23** [reply]
No, no no... sorry. Meant crossed postings. We should keep the xml stuff here, not in the alias PR... :)
I agree.. my goal for today is to look over the arches changes, and get the full import/export worked out. Then like you said, I think we can do all models next.
Need to figure out something with my VS... acting stupid after I pulled all the recent changes last night and now it "thinks" they are my changes.

**Reply by AzGilrock** [reply]
I used to use GUI's and such to work with Git but after getting on a project at work where I had to learn everything from the command line I like doing everything that way now.  I feel like I have better control over what's going on when I do command line Git.  Like if you were working in a branch I'd do:
1)  Checkin all your changes locallly for your branch: 
      git add *
      git commit -m "add comment here"
2) Update your master branch
      git fetch           (this pulls all remote information for the repo)
      git checkout master
      git pull
3)  Switch back to your branch and merge with master
      git checkout your_branch
      git merge master      (then resolve any files that say Conflict in the editor)

I do lots of "git status" commands to check where things are at along the way.

**Comment by cybercop23** [comment]
#4529

**Comment by cybercop23** [comment]
So currently CandyCanes export as a Custom Model and there's no option for xLights Model. Any ideas why and more importantly with this change, can/should we change it back to native? IDK that I want to jump right into the custom model export - was trying to work down the list and just do CC next.
@keithsw1111 @computergeek1507 @AzGilrock @dkulp ?

**Reply by cybercop23** [reply]
It seems to auto create subbuffer submodels to handle the way they are layed out, even if no alternate node is defied.
However the RGBEffects file keeps it as CC.. Going to change it/assume we can keep it as native model on exports too.

**Comment by cybercop23** [comment]
@AzGilrock as part of this change, should we 

1. retire wreath and convert all to cirlce (wreath is a special case for a circle)
2. change tree to only have one model, vs 3; 360, flat, ribbon?

**Reply by dkulp** [reply]
Wreath is pretty much a circle with a default buffer style of "Per Preview".   It likely could be retired, but mapping existing sequences might be hard.    

The three tree styles don't affect rendering at all, just how it appears in the previews (and thus the Per Preview buffer styles).   That said, I'm not sure removing them makes sense as they have specific use cases they model, particularly for folks coming from LOR and LOR "kits".

The one thing that we DO need to get rid of is the "Whole House Model".    No one has been able to create or update one of those for at least 5 years now so I'd assume they are long gone.

**Reply by cybercop23** [reply]
"The one thing that we DO need to get rid of is the "Whole House Model". No one has been able to create or update one of those for at least 5 years now so I'd assume they are long gone." - didn't even know there was such a thing... 

Well... lets kill it.... time to move on...
/*
 The WholeHouseModel is an ancient and long since deprecated class to represent a group of models.  There
 is currently no way to create one of these.  The class is only here to allow loading of older setups that
 have one
 */

although i like the "represent a group of models".. like we've been looking to be able to "tie" tree and star together.... hmmmm

**Reply by dkulp** [reply]
The "Whole House Model" predates supporting groups.  Originally, it was thought that dynamically creating the "Per Preview" style buffer for a large number of models would be very expensive (it likely was) so we had a specific "Whole House Model" that was a list of models that the user specifically had to click "Update" to create the buffer and it ended up being saved very similar to how a custom model is saved (x/y coord for each node, etc..).   Once the concept of Groups was added and we were able to efficiently and quickly generate the buffer layouts from the models, the Whole House Model didn't make any sense to keep around so we disabled the ability to create/update them, but left the code to load them for compatibility.   Basically, they were the equivalent of a Group now, but only supported Per Preview buffer style and only updated on demand by the user.   They are long since obsolete.

**Reply by dkulp** [reply]
I just unpacked 33,192 crash logs from 2022 and 2023 and did a grep on the rgbeffects.xml  files for 'DisplayAs="Wreath"' and found 5,899 unique wreaths.      Thus, it's definitely still used.

**Reply by cybercop23** [reply]
Maybe at the least we should remove the model from the menu so only existing ones can be copied/paste but no new ones created and come back later and see if/how we can convert them?

**Reply by dkulp** [reply]
Interesting... there were 17 crash logs with a WholeHouse model defined....    Kind of crazy.

**Reply by cybercop23** [reply]
what versions of xlights? <2020?

**Reply by dkulp** [reply]
No... I only unpacked the crash logs for 2022 and 2023 (and now 2024) versions.   There were only two in the 2023 versions, and those were in 2023.01.  None in 2024.   So over a year since we've seen a crash with a WholeHouse model in the rgbeffects.

**Comment by AzGilrock** [comment]
Yeah I had a flat ribbon tree for like 6 years.  I've retired my prop but some people might still be doing flat trees.

**Reply by cybercop23** [reply]
yes, yes... just saying to change to only have one model vs the 3 now and and the type as an attribute...
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/d74beb3d-dc64-416e-ba6e-3f92b37d2a33)

**Reply by AzGilrock** [reply]
Yes it could probably be converted to an attribute.  The ModelManager would need to have upgrade code to convert the model for existing sequences.  Or really layouts not sequences.

**Comment by cybercop23** [comment]
@AzGilrock Arch, CC and Circles all done (may need some attribure review) and all models added, visitors added and just need to create the Deserializers for the rest. But before I do that, I'd like to be able to get the submodels, aliases, groups to pull real values. I'll draft checxk-in what I have so you can see, but I just need a hand to get to the darn models. When you look at the code, you'll be able to see what I mean.

Updating Draft PR

**Reply by AzGilrock** [reply]
Ok let me know when to look.  BTW the MultiPointModel is one that is a work-in-progress so currently I don't enable the button for the user to create one.  Technically you can edit an rgbeffects file and convert a PolyLine to the MultiPoint and see it work in the released version.  Its basically a PolyLine where every corner is a pixel.  It will be able to be used to create more accurate custom models since the positions will be floating point and not limited to grid locations.  I was working on making an editor and eventually have it like David Peace's web tool for model editing and I was going to have some model auto-generation features.

**Reply by cybercop23** [reply]
Just checked them in... i ignored all headers that i added visitors thinking it will not send them, but it did, ignore those.
so really just the baseobjectvisitor has all defs and the xmlserializer

**Comment by cybercop23** [comment]
UPDATE.... (I need to get better at only keeping comments/questions in just one place - that PR is full if info/questions. - maybe I can bring them into here).

1. I have uploaded the latest version. It can export and import "compressed" CM with depths. Right now it has a separate attribute CustomModel2.0 since we need to create an "upgrade" function when we change the versionnumber. 
RGBEffects will keep and save the compressed model if exported/imported but doesn't yet touch existing models.
* Model Data notebook panel is blank since it can't read the compressed format and I really don't want to add an expansion unless it is truly necessary and even then only use it for that one item.

2. I've started to look at the other objects, currently workign on views, but I'm stuck, I don't know how to integrate them since they aren't a type Model and the current Factory only deals with Models. Need some help to continue adding these.

3. Maybe we can/should do them in sections. Do the models and come back for the rest. I'm affraid there will be lots of time spent to resolve conflicts. Not to mention the recently surprizing discovery of reinit time improvements.

**Comment by cybercop23** [comment]
@AzGilrock how the hell do I get to the View objects to get their values? Tried everyhing. TerrianObject (we have to change them to Terrain) all is blank, if I create my own function to pass and read an attribute like it doesn on InitModel .. I get defaults... I'm missing something.
With the exception of palettes and this I have all of the other non-model related visitor info done.

**Reply by AzGilrock** [reply]
Yeah saw the misspelling awhile back but figured didn't care enough to go to the trouble lol

**Reply by AzGilrock** [reply]
I'll look at the view objects when I'm on my computer

**Comment by cybercop23** [comment]
> Model Data notebook panel is blank since it can't read the compressed format and I really don't want to add an expansion unless it is truly necessary and even then only use it for that one item.
Fixed.. would have been a good idea to actually look at the code and see what it does.... (x,y) -> value... hehehe... parsed the old format only to put it in the format we have now.
